crates/sui-framework/sources/safe.move

```diff
@@ -0,0 +1,214 @@
+        let capability = Capability {
+            id: object::new(ctx),
+            safe_id: object::uid_to_inner(&safe.id),
+            nonce: safe.next_nonce,
+            amount: withdraw_amount,
+        };
+
+        bitmap_expand(&mut safe.permissions_bitmap, safe.next_nonce);
+        bitmap_set(&mut safe.permissions_bitmap, safe.next_nonce, true);
+
+        safe.next_nonce = safe.next_nonce + 1;
+        capability
+    }
+
+    //////////////////////////////////////////////////////
+    /// PUBLIC FUNCTIONS
+    //////////////////////////////////////////////////////
+
+    /// Wrap a coin around a safe.
+    /// a trusted party (or smart contract) to transfer the object out.
+    public fun create<T>(balance: Balance<T>, ctx: &mut TxContext) {
+        let box = Safe {
+            id: object::new(ctx),
+            balance,
+            owner: sender(ctx),
+            next_nonce: 0,
+            permissions_bitmap: vector::empty(),
+        };
+        transfer::share_object(box)
+    }
+
+    /// Wraps an object with type T into a lootbox, and returns a transfer capability. Useful for creating a safe and
+    /// providing permissions to an external party/smart contract within one transaction.
+    public fun create_with_capability<T>(balance: Balance<T>, capability_amount: u64, ctx: &mut TxContext): Capability<T> {
+        let safe = Safe {
+            id: object::new(ctx),
+            balance,
+            owner: sender(ctx),
+            next_nonce: 0,
+            permissions_bitmap: vector::empty(),
+        };
+
+        let capability = Capability {
+            id: object::new(ctx),
+            safe_id: object::uid_to_inner(&safe.id),
+            amount: capability_amount,
+            nonce: 0,
+        };
+
+        transfer::share_object(safe);
+        capability
+    }
+
+    /// Withdraw coins from the safe as an owner.
+    public fun withdraw_as_owner<T>(safe: &mut Safe<T>, withdraw_amount: u64, ctx: &mut TxContext): Balance<T> {
+        // Ensures that only the owner can withdraw from the safe.
+        assert!(sender(ctx) == safe.owner, 0);
+        balance::split(&mut safe.balance, withdraw_amount)
+    }
+
+    /// CAPABILITY OWNER ONLY
+    /// Withdraw coins from the safe as a capability holder.
+    public fun withdraw_as_capability_holder<T>(safe: &mut Safe<T>, withdraw_amount: u64, capability: &mut Capability<T>): Balance<T> {
+        // Check that the capability pointer matches the UID of the lootbox. Abort if that is not the case.
+        assert!(object::uid_to_inner(&safe.id) == capability.safe_id, 0);
+
+        // Update the funds. Note that this will revert if there aren't sufficient funds.
+        capability.amount = capability.amount - withdraw_amount;
+
+        balance::split(&mut safe.balance, withdraw_amount)
+    }
+
+    /// Safe Owner Only
+    public fun revoke_capability<T>(safe: &mut Safe<T>, capability: &Capability<T>, ctx: &TxContext) {
```

Comment:
I think this probably wants to take the `nonce` as input. Otherwise, a malicious `Capability` holder can prevent you from revoking their capability by making it a single-owner object, or by wrapping it.

Reply:
I think this (in turn) means that either the `Safe` must remember the `amount` of each capability it issued (otherwise, it won't know what to subtract during revocation), or must only support a `revoke_all` API.

Reply:
Yes, I've just realized that this is a problem - a fix was just pushed in.

Reply:
Why does the Safe have to remember the amount of each capability? Currently revoking just completely turns off the capability.

Reply:
However, another thing we might want to consider is whether it is desirable for the author to be able to change the amounts which they allow capabilities to use. In this case, we could instead store a vector of u64 amounts that represent how much each capability could take out.